text/0000-router-service-computed-query-params.md

```diff
@@ -0,0 +1,324 @@
+- Start Date: 2018-09-21
+- RFC PR: [https://github.com/emberjs/rfcs/pull/380](https://github.com/emberjs/rfcs/pull/380)
+- Ember Issue: (leave this empty)
+
+# Add `queryParams` to the router service
+
+## Summary
+
+This RFC proposes a new primitive API change to the `RouterService` to allow access to query params from anywhere.
+
+Access to query params is currently restricted to the controller, and subsequently, the corresponding route.
+This results in some limitations with which users may consume query param data.
+By exposing query params on the [RouterService](https://api.emberjs.com/ember/release/classes/RouterService), users will be able to easily access the query params from deep down a component tree, removing the need to pass query param related data and actions down many layers of components.
+
+
+
+## Motivation
+
+Modern SPA concepts have converged on the idea that query params should be easily accessible -- independent from the object responsible for handling the route.
+Like with the [RouterService](https://github.com/emberjs/rfcs/blob/master/text/0095-router-service.md),
+it is common to have a need to perform routing behavior from deep down a component tree.
+Additionally, the current query params implementation feels very verbose for "just wanting to access a property" and has been frustrating to have to explain awkward behavior when on-boarding new devs who may be unfamiliar with Ember.
+
+Accessing data within the url **should feel easy**.
+
+**What's wrong with the existing query params?**
+- For all but one use case, controllers can be avoided. Query Params force controllers into existence for those who are trying to avoid them. 
+- The caching mechanism is persistent beyond just child routes. _Any_ time a route with query params is re-visited, the query params values will be restored. This can be useful for those who are needing this behavior, but for those who want to manage queryParams via transition / navigations, they'd need to set up query param resets on enter/exit of routes, link-to's and transitions -- the query params implementation becomes more of an obstacle than a feature.
+
+## Detailed Design
+
+### Accessing Query Params
+
+```ts
+import Component from "@glimmer/component";
+import { inject as service } from '@ember/service';
+
+export default class Pagination extends Component {
+  @service router;
+
+  get currentPage() {
+    const { page } = this.router.queryParams;
```

Comment:
Why do we need to expand the surface area of the Router Service? There is already `currentRoute.queryParams` on the service.

Reply:
I think the intent is that queryParams just.. _are_ and aren't tied to a specific route.
`currentRoute.queryParams`, at least to me, implies that other routes could have different set's of query params.

That said, I don't see a reason why `currentRoute.queryParams` couldn't be the place to get query params -- I just think it implies unwanted things.

Reply:
Maybe some third party cacheing addon could provide different query params per route?